homeassistant/components/rainbird/__init__.py

```diff
@@ -3,40 +3,82 @@
 
 import voluptuous as vol
 
+from homeassistant.components import binary_sensor, sensor, switch
+from homeassistant.const import (
+    CONF_FRIENDLY_NAME,
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_TRIGGER_TIME,
+)
+from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
-from homeassistant.const import CONF_HOST, CONF_PASSWORD
+
+SUPPORTED_PLATFORMS = [switch.DOMAIN, sensor.DOMAIN, binary_sensor.DOMAIN]
 
 _LOGGER = logging.getLogger(__name__)
 
+RAINBIRD_CONTROLLER = "controller"
 DATA_RAINBIRD = "rainbird"
 DOMAIN = "rainbird"
 
-CONFIG_SCHEMA = vol.Schema(
+SENSOR_TYPE_RAINDELAY = "raindelay"
+SENSOR_TYPE_RAINSENSOR = "rainsensor"
+# sensor_type [ description, unit, icon ]
+SENSOR_TYPES = {
+    SENSOR_TYPE_RAINSENSOR: ["Rainsensor", None, "mdi:water"],
+    SENSOR_TYPE_RAINDELAY: ["Raindelay", None, "mdi:water-off"],
+}
+
+ZONE_SCHEMA = vol.Schema(
     {
-        DOMAIN: vol.Schema(
-            {vol.Required(CONF_HOST): cv.string, vol.Required(CONF_PASSWORD): cv.string}
-        )
-    },
+        vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+        vol.Optional(CONF_TRIGGER_TIME): cv.positive_int,
+        vol.Optional(CONF_SCAN_INTERVAL): cv.positive_int,
+    }
+)
+CONTROLLER_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_TRIGGER_TIME): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL): cv.string,
+        vol.Optional("zones"): vol.Schema({cv.positive_int: ZONE_SCHEMA}),
+    }
+)
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.Any(cv.ensure_list(CONTROLLER_SCHEMA), CONTROLLER_SCHEMA))},
     extra=vol.ALLOW_EXTRA,
 )
 
 
 def setup(hass, config):
     """Set up the Rain Bird component."""
-    conf = config[DOMAIN]
-    server = conf.get(CONF_HOST)
-    password = conf.get(CONF_PASSWORD)
 
-    from pyrainbird import RainbirdController
+    hass.data[DATA_RAINBIRD] = list()
+    if isinstance(config[DOMAIN], list):
```

Comment:
Remove all the checks, and just iterate the list of controllers. The config validation will make sure it's always a list.

Reply:
ok

Reply:
Just to be clear, with the updated config, the user can provide a single controller and a list of controllers. The config validation will turn both cases into a list of controllers. That's what `cv.ensure_list` does.